os/lib/cbor.c

```diff
@@ -0,0 +1,434 @@
+{
+  write_first_byte(state, CBOR_SIMPLE_VALUE_UNDEFINED);
+}
+/*---------------------------------------------------------------------------*/
+void
+cbor_write_bool(cbor_writer_state_t *state, bool boolean)
+{
+  write_first_byte(state,
+                   boolean ? CBOR_SIMPLE_VALUE_TRUE : CBOR_SIMPLE_VALUE_FALSE);
+}
+/*---------------------------------------------------------------------------*/
+static void
+generic_open(cbor_writer_state_t *state, cbor_major_type_t major_type)
+{
+  if(!state->nesting_depth) {
+    cbor_break_writer(state);
+    return;
+  }
+  write_first_byte(state, major_type);
+  state->records[--state->nesting_depth].start = state->buffer;
+  state->records[state->nesting_depth].objects = 0;
+}
+/*---------------------------------------------------------------------------*/
+static void
+generic_close(cbor_writer_state_t *state, size_t value)
+{
+  insert_unsigned(state, state->records[state->nesting_depth].start, value);
+  state->nesting_depth++;
+}
+/*---------------------------------------------------------------------------*/
+void
+cbor_open_data(cbor_writer_state_t *state)
+{
+  generic_open(state, CBOR_MAJOR_TYPE_BYTE_STRING);
+}
+/*---------------------------------------------------------------------------*/
+void
+cbor_close_data(cbor_writer_state_t *state)
+{
+  if(state->nesting_depth == CBOR_MAX_NESTING) {
+    cbor_break_writer(state);
+    return;
+  }
+  generic_close(state,
+                state->buffer - state->records[state->nesting_depth].start);
+}
+/*---------------------------------------------------------------------------*/
+void
+cbor_open_array(cbor_writer_state_t *state)
+{
+  generic_open(state, CBOR_MAJOR_TYPE_ARRAY);
+}
+/*---------------------------------------------------------------------------*/
+void
+cbor_close_array(cbor_writer_state_t *state)
+{
+  if(state->nesting_depth == CBOR_MAX_NESTING) {
+    cbor_break_writer(state);
+    return;
+  }
+  generic_close(state, state->records[state->nesting_depth].objects);
+}
+/*---------------------------------------------------------------------------*/
+void
+cbor_open_map(cbor_writer_state_t *state)
+{
+  generic_open(state, CBOR_MAJOR_TYPE_MAP);
+}
+/*---------------------------------------------------------------------------*/
+void
+cbor_close_map(cbor_writer_state_t *state)
+{
+  if((state->nesting_depth == CBOR_MAX_NESTING)
+     || (state->records[state->nesting_depth].objects & 1)) {
+    cbor_break_writer(state);
+    return;
+  }
+  generic_close(state, state->records[state->nesting_depth].objects >> 1);
+}
+/*---------------------------------------------------------------------------*/
+void
+cbor_init_reader(cbor_reader_state_t *state,
+                 const uint8_t *cbor, size_t cbor_size)
+{
+  state->cbor = cbor;
+  state->cbor_size = cbor_size;
+}
+/*---------------------------------------------------------------------------*/
+cbor_major_type_t
+cbor_read_next(cbor_reader_state_t *state)
```

Comment:
All read functions except for this consume data. What do you think about naming it `cbor_peek_next()` or similar?

Reply:
Done